flare_portal/experiments/forms.py

```diff
@@ -68,6 +71,71 @@ def save(self, *, experiment: Experiment) -> None:
         )
 
 
+class ParticipantUploadForm(forms.Form):
+    import_file = forms.FileField(
+        validators=[FileExtensionValidator(["csv"])],
+    )
+
+    @staticmethod
+    def open_csv(field):
+        # Open uploaded file
+        data = io.StringIO(field.read().decode("utf-8"))
+        return csv.DictReader(data)
+
+    def clean(self):
+        # Don't continue if no upload
+        file = self.cleaned_data.get("import_file")
+        if file is None:
+            return
+
+        # Open uploaded file
+        data = io.StringIO(file.read().decode("utf-8"))
+        upload = csv.DictReader(data)
+
+        # Build a list of ID's
+        pids = []
+        for row in upload:
+            if pid := row.get("pid"):
+                if len(pid) <= 24 and pid not in pids:
+                    pids.append(pid)
+                else:
+                    self.add_error(
+                        "import_file",
+                        "Participant ID's must be less than 25 characters.",
+                    )
+
+        # Validate and remove any ID's that already exist
+        row_count = len(pids)
+        existing_participants = Participant.objects.filter(participant_id__in=pids)
+        for participant in existing_participants:
+            pids.remove(participant.participant_id)
```

Comment:
Suggestion: General implementation of this is OK so stick to it if you can't be bothered. But I think it would make sense to use sets to ensure there are no duplicate IDs. It also makes it easier to remove elements all in one go given the set of existing participant ids.